File: mobile/MultiTenantEcommerce.Maui/Services/TokenStorageService.cs
using Microsoft.Maui.Storage;

namespace MultiTenantEcommerce.Maui.Services;

public class TokenStorageService
{
    private const string AccessTokenKey = "mt_access";
    private const string RefreshTokenKey = "mt_refresh";
    private const string ExpiresKey = "mt_expires";
    private const string TenantKey = "mt_tenant";

    public async Task SetAsync(string accessToken, string refreshToken, DateTime expiresAt, string tenant)
    {
        await SecureStorage.SetAsync(AccessTokenKey, accessToken);
        await SecureStorage.SetAsync(RefreshTokenKey, refreshToken);
        await SecureStorage.SetAsync(ExpiresKey, expiresAt.ToString("O"));
        await SecureStorage.SetAsync(TenantKey, tenant);
    }

    public async Task<TokenBundle?> GetAsync()
    {
        var accessToken = await SecureStorage.GetAsync(AccessTokenKey);
        var refreshToken = await SecureStorage.GetAsync(RefreshTokenKey);
        var expiresRaw = await SecureStorage.GetAsync(ExpiresKey);
        var tenant = await SecureStorage.GetAsync(TenantKey);

        if (string.IsNullOrEmpty(accessToken) || string.IsNullOrEmpty(refreshToken) || string.IsNullOrEmpty(expiresRaw) || string.IsNullOrEmpty(tenant))
        {
            return null;
        }

        return new TokenBundle(accessToken, refreshToken, DateTime.Parse(expiresRaw), tenant);
    }

    public void Clear()
    {
        SecureStorage.Remove(AccessTokenKey);
        SecureStorage.Remove(RefreshTokenKey);
        SecureStorage.Remove(ExpiresKey);
        SecureStorage.Remove(TenantKey);
    }
}

public record TokenBundle(string AccessToken, string RefreshToken, DateTime ExpiresAt, string Tenant);

File: frontend/angular/src/app/core/token-storage.service.ts
import { Injectable } from '@angular/core';

export interface StoredTokens {
  accessToken: string;
  refreshToken: string;
  expiresAt: Date;
  tenantHeader?: string;
}

@Injectable({ providedIn: 'root' })
export class TokenStorageService {
  private readonly accessTokenKey = 'mt_access';
  private readonly refreshTokenKey = 'mt_refresh';
  private readonly expiresKey = 'mt_expires';
  private readonly tenantKey = 'mt_tenant';

  setTokens(tokens: StoredTokens): void {
    localStorage.setItem(this.accessTokenKey, tokens.accessToken);
    localStorage.setItem(this.refreshTokenKey, tokens.refreshToken);
    localStorage.setItem(this.expiresKey, tokens.expiresAt.toISOString());
    if (tokens.tenantHeader) {
      localStorage.setItem(this.tenantKey, tokens.tenantHeader);
    } else {
      localStorage.removeItem(this.tenantKey);
    }
  }

  getTokens(): StoredTokens | null {
    const accessToken = localStorage.getItem(this.accessTokenKey);
    const refreshToken = localStorage.getItem(this.refreshTokenKey);
    const expiresAtRaw = localStorage.getItem(this.expiresKey);
    const tenant = localStorage.getItem(this.tenantKey) ?? undefined;

    if (!accessToken || !refreshToken || !expiresAtRaw) {
      return null;
    }

    return {
      accessToken,
      refreshToken,
      expiresAt: new Date(expiresAtRaw),
      tenantHeader: tenant
    };
  }

  clear(): void {
    localStorage.removeItem(this.accessTokenKey);
    localStorage.removeItem(this.refreshTokenKey);
    localStorage.removeItem(this.expiresKey);
    localStorage.removeItem(this.tenantKey);
  }
}
